File: 04-stack/00853-Car_Fleet/00853-car_fleet.cpp
// 853: Car Fleet
// https://leetcode.com/problems/car-fleet/


#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>

using namespace std;


class Solution {
    class car {
    public:
        car(int p, int s): p(p), s(s) {};
        int p;
        int s;
    };
    
public:
    // SOLUTION
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        vector<car> cars;
        int n = position.size();
        for (int i=0; i<n; i++)
            cars.emplace_back(position[i], speed[i]);

        sort(cars.begin(), cars.end(), [](const car& a, car& b) {return a.p < b.p;});

        stack<float> s;
        for (int i=0; i<n; i++) {
            float time = (target - cars[i].p) / (float)cars[i].s;
            while (!s.empty() && time >= s.top()) s.pop();
            s.push(time);
        }

        return s.size();
    }
};


int main() {
    Solution o;

    // INPUT
    int target = 12;
    vector<int> position = {10,8,0,5,3};
    vector<int> speed = {2,4,1,1,3};

    // OUTPUT
    auto result = o.carFleet(target, position, speed);
    cout<<result<<endl;

    return 0;
}

File: 04-stack/00853-Car_Fleet/00853-car_fleet.java
// 853: Car Fleet
// https://leetcode.com/problems/car-fleet/


import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Stack;


class Solution {
    static public class car {
        public car(int p, int s) {this.p = p; this.s = s;};
        public int p;
        public int s;
    }

    // SOLUTION
    public int carFleet(int target, int[] position, int[] speed) {
        ArrayList<car> cars = new ArrayList<>();
        int n = position.length;
        for (int i=0; i<n; i++)
            cars.add(new car(position[i], speed[i]));

        Collections.sort(cars, new Comparator<car>() {
            public int compare(car a, car b) {
                return a.p - b.p;
            }
        });

        Stack<Double> s = new Stack<>();
        for (int i=0; i<n; i++) {
            double time = (target - cars.get(i).p) / (double)cars.get(i).s;
            while (!s.isEmpty() && time >= s.peek()) s.pop();
            s.push(time);
        }

        return s.size();
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int target = 12;
        int[] position = {10,8,0,5,3};
        int[] speed = {2,4,1,1,3};

        // OUTPUT
        var result = o.carFleet(target, position, speed);
        System.out.println(result);
    }
}